src/librustc_mir/build/block.rs

```diff
@@ -121,12 +122,15 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                                 let scope = (init_scope, source_info);
                                 this.in_scope(scope, lint_level, block, move |this| {
                                     // FIXME #30046                             ^~~~
-                                    this.expr_into_pattern(block, pattern, init)
+                                    this.expr_into_pattern(block, ty, pattern, init)
                                 })
                             }));
                     } else {
                         this.visit_bindings(&pattern, &mut |this, _, _, node, span, _| {
                             this.storage_live_binding(block, node, span);
+                            if let Some(ty) = ty {
```

Comment:
So, this won't work for something like `let (a, b): (u32, u32) = (22, 44);` -- it will assert that both `a` and `b` have type `(u32, u32)`, I believe.

Reply:
I'm not sure how most elegantly to handle this case right now -- one way would be to track the `ty` as we visit the bindings, so that we can destructure it appropriately. Another would be to "build back up" the value from the types of the bindings. I don't like either one, unfortunately :(

Reply:
(I wonder if we can push the problem to typeck, actually, since it already has to do "destructure" the type hint, I think?)